silkrpc/context_pool.hpp

```diff
@@ -31,37 +31,253 @@
+    inline static const uint32_t kSleep1EveryHowManyTimes{20};
+
+    uint32_t counter_{0};
 };
 
-std::ostream& operator<<(std::ostream& out, const Context& c);
+struct BusySpinWaitStrategy : public WaitStrategy {
+    void wait_once(uint32_t /*executed_count*/) override {
+    }
+};
+
+enum class WaitMode {
+    blocking,
+    sleeping,
+    yielding,
+    spin_wait,
+    busy_spin
+};
+
+std::unique_ptr<WaitStrategy> make_wait_strategy(WaitMode wait_mode);
 
 using ChannelFactory = std::function<std::shared_ptr<grpc::Channel>()>;
 
+//! Asynchronous client scheduler running an execution loop.
+class Context {
+  public:
+    explicit Context(ChannelFactory create_channel, std::shared_ptr<BlockCache> block_cache, WaitMode wait_mode = WaitMode::blocking);
+
+    asio::io_context* io_context() const noexcept { return io_context_.get(); }
+    grpc::CompletionQueue* grpc_queue() const noexcept { return queue_.get(); }
+    silkworm::rpc::CompletionEndPoint* rpc_end_point() noexcept { return rpc_end_point_.get(); }
+    std::unique_ptr<ethdb::Database>& database() noexcept { return database_; }
+    std::unique_ptr<ethbackend::BackEnd>& backend() noexcept { return backend_; }
+    std::unique_ptr<txpool::Miner>& miner() noexcept { return miner_; }
+    std::unique_ptr<txpool::TransactionPool>& tx_pool() noexcept { return tx_pool_; }
+    std::shared_ptr<BlockCache>& block_cache() noexcept { return block_cache_; }
+
+    //! Execute the scheduler loop until stopped.
+    void execute_loop();
+
+    //! Stop the execution loop.
+    void stop();
+
+  private:
+    //! Execute single-threaded loop until stopped.
+    void execute_loop_single_threaded();
+
+    //! Execute single-threaded loop until stopped.
+    void execute_loop_double_threaded();
+
+    //! The asynchronous event loop scheduler.
+    std::shared_ptr<asio::io_context> io_context_;
+
+    //! The work-tracking executor that keep the scheduler running.
+    asio::execution::any_executor<> work_;
+
+    std::unique_ptr<grpc::CompletionQueue> queue_;
+    std::unique_ptr<silkworm::rpc::CompletionEndPoint> rpc_end_point_;
+    std::unique_ptr<ethdb::Database> database_;
+    std::unique_ptr<ethbackend::BackEnd> backend_;
+    std::unique_ptr<txpool::Miner> miner_;
+    std::unique_ptr<txpool::TransactionPool> tx_pool_;
+    std::shared_ptr<BlockCache> block_cache_;
+    WaitMode wait_mode_;
+};
+
+std::ostream& operator<<(std::ostream& out, Context& c);
+
 class ContextPool {
 public:
-    explicit ContextPool(std::size_t pool_size, ChannelFactory create_channel);
+    explicit ContextPool(std::size_t pool_size, ChannelFactory create_channel, WaitMode wait_mode = WaitMode::blocking);
+    ~ContextPool();
 
     ContextPool(const ContextPool&) = delete;
     ContextPool& operator=(const ContextPool&) = delete;
 
-    void run();
+    void start();
+
+    void join();
 
     void stop();
 
+    void run();
+
     Context& get_context();
```

Comment:
instead of get_context() could be useful rename something that explain that we are selecting a context (select_context()) with this name seems only a get method

Reply:
Good point. I think `select_context` or `next_context` are better names than simply `get_context`